ballet/validation/entropy.py

```diff
@@ -0,0 +1,165 @@
+import numpy as np
+from scipy.special import digamma
+from sklearn.neighbors import NearestNeighbors
+
+from ballet.util import asarray2d
+
+NUM_NEIGHBORS = 3  # Used in the sklearn mutual information function
+
+
+def calculate_disc_entropy(X):
+    # An exact calculation of the dataset entropy, using empirical probability
+    # H = sum(p_i * log2(p_i))
+    n_samples, _ = X.shape
+    _, counts = np.unique(X, axis=0, return_counts=True)
+    empirical_p = counts * 1.0 / n_samples
+    log_p = np.log(empirical_p)
+    entropy = -np.sum(np.multiply(empirical_p, log_p))
+    return entropy
+
+
+def estimate_cont_entropy(X, epsilon=None):
+    """
+    Based off the Kraskov Estimator for Shannon Entropy
+    Implementation based off summary here:
+    https://journals.aps.org/pre/pdf/10.1103/PhysRevE.69.066138
+    Calculating eq. 22 minus the last term (cancels out)
+    """
+    n_samples, n_features = X.shape
+    if n_samples <= 1:
+        return 0
+    nn = NearestNeighbors(
+        metric='chebyshev',
+        n_neighbors=NUM_NEIGHBORS,
+        algorithm='kd_tree')
+    nn.fit(X)
+    if epsilon is None:
+        # If epsilon is not provided, revert to the Kozachenko Estimator
+        n_neighbors = NUM_NEIGHBORS
+        radius = None
+        # While we have non-zero radii, calculate for a larger k
+        # Potentially expensive
+        while not np.all(radius) and n_neighbors < n_samples:
+            distances, _ = nn.kneighbors(
+                n_neighbors=n_neighbors, return_distance=True)
+            radius = distances[:, -1]
+            n_neighbors += 1
+        if n_neighbors == n_samples:
+            # This case only happens if all samples are the same
+            # e.g. this isn't a continuous sample...
+            raise ValueError('Should not have discrete column to estimate')
+        return -digamma(n_neighbors) + digamma(n_samples) + \
+            n_features * np.mean(np.log(2 * radius))
+    else:
+        ind = nn.radius_neighbors(
+            radius=epsilon.ravel(),
+            return_distance=False)
+        nx = np.array([i.size for i in ind])
+        return - np.mean(digamma(nx + 1)) + digamma(n_samples)
+
+
+def _is_column_discrete(col):
+    # Stand-in method to figure out if column is discrete
```

Comment:
add the heuristics for integer values and integer-adjacent floating point values